components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java

```diff
@@ -726,29 +859,101 @@ then any unregistered scopes(excluding internal scopes
+                        IdentityTenantUtil.getTenantId(params.getTenantDomain()));
+                if (existingUserConsent != null) {
+                    if (CollectionUtils.isNotEmpty(existingUserConsent.getApprovedScopes())) {
+                        allowedOAuthScopes.removeAll(existingUserConsent.getApprovedScopes());
+                    }
+                    // ToDo: Remove disapproved scopes from allowed scopes after implementing selected scope approval.
+                }
+            }
+            if (CollectionUtils.isNotEmpty(allowedOAuthScopes)) {
+                params.setConsentRequiredScopes(new HashSet<>(allowedOAuthScopes));
+                return allowedOAuthScopes.stream().collect(Collectors.joining(" ")).trim();
+            } else {
+                return StringUtils.EMPTY;
+            }
+        } catch (IdentityOAuth2ScopeConsentException e) {
+            throw new OAuthSystemException("Error occurred while retrieving user consents OAuth scopes.");
+        }
+    }
+
+    private static String getUserIdOfAuthenticatedUser(AuthenticatedUser user) throws OAuthSystemException {
+
+        String userId;
+        if (!user.isFederatedUser()) {
+            try {
+                userId = FrameworkUtils.resolveUserIdFromUsername(
+                        IdentityTenantUtil.getTenantId(user.getTenantDomain()),
+                        user.getUserStoreDomain(), user.getUserName());
+            } catch (UserSessionException e) {
+                throw new OAuthSystemException("Error occurred while resolving user id from authenticated user with " +
+                        "username : " + user.getUserName(), e);
+            }
+        } else {
+            try {
+                IdentityProvider idp = idpManager.getIdPByName(user.getFederatedIdPName(), user.getTenantDomain());
+                int idpId = Integer.parseInt(idp.getId());
+                userId = UserSessionStore.getInstance().getUserId(user.getUserName(),
+                        IdentityTenantUtil.getTenantId(user.getTenantDomain()), null, idpId);
+            } catch (NumberFormatException | IdentityProviderManagementException e) {
+                throw new OAuthSystemException("Error occurred while retrieving IDP id for the federated IDP : " +
+                        user.getFederatedIdPName() + " in tenant : " + user.getTenantDomain(), e);
+            } catch (UserSessionException e) {
+                throw new OAuthSystemException("Error occurred while retrieving user id of the federated user " +
+                        "for IDP : " + user.getFederatedIdPName() + " in tenant : " + user.getTenantDomain());
+            }
+        }
+        return userId;
+    }
+
+    private static String getAppIdFromClientId(String clientId) throws OAuthSystemException {
+
+        try {
+            ServiceProvider sp = OAuth2Util.getServiceProvider(clientId);
+            if (sp != null) {
```

Comment:
if (sp != null) {
                return sp.getApplicationResourceId();
            } 
 throw new OAuthSystemException("Unable to find an service provider with client Id : " + clientId);